runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1086 @@
+		#[pallet::weight((1_000, DispatchClass::Operational))]
+		pub fn assign_temp_parachain_slot(
+			origin: OriginFor<T>,
+			id: ParaId,
+			lease_period_start: SlotLeasePeriodStart,
+		) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin)?;
+
+			let manager = T::Registrar::manager_of(id).ok_or(Error::<T>::ParaDoesntExist)?;
+
+			ensure!(T::Registrar::is_parathread(id), Error::<T>::NotParathread,);
+
+			ensure!(
+				!Self::has_permanent_slot(id) && !Self::has_temporary_slot(id),
+				Error::<T>::SlotAlreadyAssigned
+			);
+
+			let current_lease_period: T::BlockNumber = Self::lease_period_index();
+			ensure!(
+				!T::Leaser::already_leased(
+					id,
+					current_lease_period,
+					// Check current lease & next one
+					current_lease_period.saturating_add(
+						T::BlockNumber::from(2u32)
+							.saturating_mul(T::TemporarySlotLeasePeriodLength::get().into())
+					)
+				),
+				Error::<T>::OngoingLeaseExists
+			);
+
+			TemporarySlots::<T>::insert(
+				id,
+				Some(ParachainTemporarySlot {
+					manager,
+					period_begin: match lease_period_start {
+						SlotLeasePeriodStart::Current => current_lease_period,
+						SlotLeasePeriodStart::Next => current_lease_period + One::one(),
+					},
+					period_count: T::TemporarySlotLeasePeriodLength::get().into(),
+					last_lease: None,
+					lease_count: 0,
+				}),
+			);
+
+			if lease_period_start == SlotLeasePeriodStart::Current {
+				Self::allocate_temporary_slot_leases(current_lease_period)?;
+			}
+
+			Self::deposit_event(Event::<T>::TemporarySlotAssigned(id));
+
+			Ok(())
+		}
+
+		/// Unassign a permanent or temporary parachain slot
+		#[pallet::weight((1_000, DispatchClass::Operational))]
+		pub fn unassign_parachain_slot(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin.clone())?;
+
+			ensure!(
+				Self::has_permanent_slot(id) || Self::has_temporary_slot(id),
+				Error::<T>::SlotNotAssigned
+			);
+
+			if PermanentSlots::<T>::contains_key(id) {
+				<PermanentSlotCount<T>>::try_mutate(|count| -> DispatchResult {
+					*count = count.checked_sub(1).ok_or(ArithmeticError::Underflow)?;
+					Ok(())
+				})?;
+				PermanentSlots::<T>::remove(id);
+			} else if TemporarySlots::<T>::contains_key(id) {
+				TemporarySlots::<T>::remove(id);
+			}
+
+			// Clean any para lease
+			Self::clear_slot_leases(origin.clone(), id)?;
+
+			// Force downgrade to parathread (if needed) before end of lease period
+			if paras::Pallet::<T>::lifecycle(id) == Some(ParaLifecycle::Parachain) {
+				runtime_parachains::schedule_parachain_downgrade::<T>(id)
+					.map_err(|_| Error::<T>::CannotDowngrade)?;
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn allocate_temporary_slot_leases(lease_period_index: LeasePeriodOf<T>) -> DispatchResult {
+		let mut active_temp_slots = 0u32;
+		let mut pending_temp_slots = Vec::new();
+		TemporarySlots::<T>::iter().for_each(|(para, maybe_slot)| {
```

Comment:
This is only constrained to the number of parathreads in the system (see note above) and could thus become very heavy.

Reply:
Add a `MaxTemporarySlots` param which constraints the nbr of temp slots.